river/cluster/denstream.py

```diff
@@ -0,0 +1,491 @@
+
+        self.centers = {i: self.clusters[i].center for i in range(self.n_clusters)}
+
+        y = self._find_closest_cluster_index(x, self.clusters)
+
+        return y
+
+
+class MicroCluster(metaclass=ABCMeta):
+    """ Micro-cluster class """
+
+    def __init__(self, x=None, timestamp=None, decaying_factor=None, current_time=None):
+
+        self.x = x
+        self.timestamp = timestamp
+        self.creation_time = timestamp
+        self.decaying_factor = decaying_factor
+        self.current_time = current_time
+
+        if x is not None and timestamp is not None:
+
+            # initial weight, initial weighted linear sum (IWLS) and initial weighted squared sum (IWSS)
+            # are calculated as described
+
+            self.N = 1
+            self.dim = len(self.x)
+            self.initial_weight = 2 ** (self.decaying_factor * self.timestamp)
+            self.LS = x
+            self.SS = {i: (x[i] * x[i]) for i in range(self.dim)}
+            self.IWLS = {
+                i: (2 ** (self.decaying_factor * self.timestamp)) * x[i]
+                for i in range(self.dim)
+            }
+            self.IWSS = {
+                i: (2 ** (self.decaying_factor * self.timestamp)) * (x[i] * x[i])
+                for i in range(self.dim)
+            }
+
+    @property
+    def weighted_linear_sum(self):
+        weighted_linear_sum = {
+            i: (2 ** (-self.current_time * self.decaying_factor)) * self.IWLS[i]
+            for i in range(self.dim)
+        }
+        return weighted_linear_sum
+
+    @property
+    def weighted_squared_sum(self):
+        weighted_squared_sum = {
+            i: (2 ** (-self.current_time * self.decaying_factor)) * self.IWSS[i]
+            for i in range(self.dim)
+        }
+        return weighted_squared_sum
+
+    @property
+    def weight(self):
+        weight = (
+            2 ** (-self.current_time * self.decaying_factor)
+        ) * self.initial_weight
+        return weight
+
+    @property
+    def center(self):
+        center = {i: self.weighted_linear_sum[i] / self.weight for i in range(self.dim)}
+        return center
+
+    @property
+    def radius(self):
+        radius = math.sqrt(
+            abs(
+                self.norm(self.weighted_squared_sum) / self.weight
+                - (self.norm(self.weighted_linear_sum) / self.weight) ** 2
+            )
+        )
+        return radius
+
+    def add(self, cluster):
+        assert self.dim == cluster.dim
+        self.N += cluster.N
+        self.initial_weight += cluster.initial_weight
+        for i in range(self.dim):
+            self.LS[i] += cluster.LS[i]
+            self.SS[i] += cluster.SS[i]
+            self.IWLS[i] += cluster.IWLS[i]
+            self.IWSS[i] += cluster.IWSS[i]
+        if self.creation_time > cluster.creation_time:
+            self.creation_time = cluster.creation_time
+
+    @staticmethod
+    def norm(x):
+        norm = 0
+        for i in range(len(x)):
```

Comment:
```python
for val in x.values():
    norm += val * val
```

Since `x` is a `dict` it means that keys can be `str`. On the other hand, even if keys are numbers they are not necessarily in order.

Reply:
Resolved!